src/rgw/rgw_data_sync.cc

```diff
@@ -3171,12 +3195,35 @@ class InitBucketFullSyncStatusCR : public RGWCoroutine {
 
       if (info.oldest_gen == 0) {
         if (check_compat) {
-          // try to convert existing per-shard incremental status for backward compatibility
-          yield call(new CheckAllBucketShardStatusIsIncremental(sc, sync_pair, num_shards, &all_incremental));
-          if (retcode < 0) {
-            return set_cr_error(retcode);
-          }
-          if (all_incremental) {
+	  yield call(new RGWSyncGetBucketInfoCR(sync_env,
+						sync_pair.source_bs.bucket,
+						&bucket_info,
+						nullptr,
+						sync_env->sync_tracer->root_node));
+
+	  yield {
+	    // use shard count from our log gen=0
+	    // try to convert existing per-shard incremental status for backward compatibility
+	    const auto& log = bucket_info.layout.logs.front();
+	    if (log.gen > 0) {
+	      ldpp_dout(dpp, -1) << "failed to init bucket shard status: "
+		"no generation zero when check compatibility" << dendl;
+	      return set_cr_error(-EINVAL);
+	    }
+	    if (log.layout.type != rgw::BucketLogType::InIndex) {
+	      ldpp_dout(dpp, -1) << "failed to init bucket shard status: "
+		"unrecognized log layout type " << log.layout.type << dendl;
+	      return set_cr_error(-EINVAL);
```

Comment:
these `EINVAL` errors shouldn't be fatal to `InitBucketFullSyncStatusCR`; they should just skip the call to `CheckAllBucketShardStatusIsIncremental` and set `all_incremental == false`

this backward-compat logic deals with the case where we haven't created a bucket-wide sync status object (which was added as part of this multisite reshard work), but do have all of the per-shard status objects. if we find that all those shards are already on incremental sync, we can initialize the bucket-wide status with `BucketSyncState::Incremental` and avoid restarting a full sync

Reply:
Done!